x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExportBulk.java

```diff
@@ -123,51 +126,43 @@ public void onFailure(Exception exception) {
         }
     }
 
-    private byte[] toBulkBytes(final MonitoringDoc doc) throws IOException {
+    private void writeDocument(MonitoringDoc doc, StreamOutput out) throws IOException {
         final XContentType xContentType = XContentType.JSON;
         final XContent xContent = xContentType.xContent();
 
         final String index = MonitoringTemplateUtils.indexName(formatter, doc.getSystem(), doc.getTimestamp());
         final String id = doc.getId();
 
-        try (BytesStreamOutput out = new BytesStreamOutput()) {
-            try (XContentBuilder builder = new XContentBuilder(xContent, out)) {
-                // Builds the bulk action metadata line
-                builder.startObject();
+        try (XContentBuilder builder = new XContentBuilder(xContent, out)) {
+            // Builds the bulk action metadata line
+            builder.startObject();
+            {
+                builder.startObject("index");
                 {
-                    builder.startObject("index");
-                    {
-                        builder.field("_index", index);
-                        if (id != null) {
-                            builder.field("_id", id);
-                        }
+                    builder.field("_index", index);
+                    if (id != null) {
+                        builder.field("_id", id);
                     }
-                    builder.endObject();
                 }
                 builder.endObject();
             }
+            builder.endObject();
+        }
 
-            // Adds action metadata line bulk separator
-            out.write(xContent.streamSeparator());
-
-            // Adds the source of the monitoring document
-            final BytesRef source = XContentHelper.toXContent(doc, xContentType, false).toBytesRef();
-            out.write(source.bytes, source.offset, source.length);
-
-            // Adds final bulk separator
-            out.write(xContent.streamSeparator());
+        // Adds action metadata line bulk separator
+        out.write(xContent.streamSeparator());
 
-            logger.trace(
-                "http exporter [{}] - added index request [index={}, id={}, monitoring data type={}]",
-                name, index, id, doc.getType()
-            );
+        // Adds the source of the monitoring document
+        try (XContentBuilder builder = new XContentBuilder(xContent, out)) {
+            doc.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        }
 
-            return BytesReference.toBytes(out.bytes());
-        } catch (Exception e) {
-            logger.warn((Supplier<?>) () -> new ParameterizedMessage("failed to render document [{}], skipping it [{}]", doc, name), e);
```

Comment:
This is hard to model with streaming IO, but to me catch and appending an empty byte array seems bogus here and not worth retaining (there's no test covering this case and we should simply make sure out documents that we have full control over serialise shouldn't we?).